File: DesignPatterns/01_GOF23/DP_examples/CSharp/03_Behavioral/Visitor/ObjectStructure.cs
namespace Behavioral.Visitor;

public class Computer : IComputerPart
{
    IComputerPart[] parts;
    public Computer()
    {
        parts = new IComputerPart[] { new Keyboard(), new Mouse(), new Monitor() };
    }
    public void Accept(IComputerPartVisitor computer)
    {
        for (int i = 0; i < parts.Length; i++)
            parts[i].Accept(computer);
        computer.Visit(this);
    }
}

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/Visitor/ObjectStructure.rs
// 具体元素：Computer (ObjectStructure)
pub struct Computer {
    pub parts: Vec<Box<dyn crate::ConcreteElement::IComputerPart>>,
}

use crate::{ConcreteElement::{IComputerPart, IComputerPartVisitor, Keyboard, Mouse, Monitor}};

// 具体元素：Computer (ObjectStructure) 实现
impl Computer {
    pub fn new() -> Self {
        Self {
            parts: vec![
                Box::new(Keyboard),
                Box::new(Mouse),
                Box::new(Monitor),
            ],
        }
    }
}

impl IComputerPart for Computer {
    fn accept(&self, visitor: &dyn IComputerPartVisitor) {
        for part in &self.parts {
            part.accept(visitor);
        }
        visitor.visit_computer(self);
    }
}
